File: go/store/redis/redis.go
// Package redis is the Redis store (REQ-ST-REDIS-1): begin, complete and abandon are each one Lua script run
// as EVALSHA with an EVAL fallback on NOSCRIPT, one hash per record, and a PEXPIRE sweeps expired hashes.
package redis

import (
	"context"
	"encoding/base64"
	"fmt"
	"strconv"
	"time"

	goredis "github.com/redis/go-redis/v9"
	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/store/internal/rowcodec"
)

// DefaultPrefix is the key prefix New uses when Options.Prefix is empty.
const DefaultPrefix = "anyonce:"

// DefaultNativeTTLGrace is added to the PEXPIRE deadline when Options.NativeTTLGrace is zero.
const DefaultNativeTTLGrace = 60 * time.Second

const inFlight = string(anyonce.StateInFlight)

// Options configure the store. Prefix defaults to DefaultPrefix. NativeTTLGrace is added to the PEXPIRE so a
// late complete from the previous fence holder still finds its hash; it defaults to DefaultNativeTTLGrace.
type Options struct {
	Prefix         string
	NativeTTLGrace time.Duration
}

// Store is the Redis implementation of anyonce.Store.
type Store struct {
	client         goredis.UniversalClient
	prefix         string
	grace          time.Duration
	beginScript    *goredis.Script
	completeScript *goredis.Script
	abandonScript  *goredis.Script
}

// New builds a store over an existing Redis client. client may be a single-node client, a sentinel-backed
// failover client, a ring or a cluster client; all satisfy goredis.UniversalClient.
func New(client goredis.UniversalClient, opts Options) *Store {
	s := &Store{
		client:         client,
		prefix:         opts.Prefix,
		grace:          opts.NativeTTLGrace,
		beginScript:    goredis.NewScript(beginLua),
		completeScript: goredis.NewScript(completeLua),
		abandonScript:  goredis.NewScript(abandonLua),
	}
	if s.prefix == "" {
		s.prefix = DefaultPrefix
	}
	if s.grace == 0 {
		s.grace = DefaultNativeTTLGrace
	}
	return s
}

// key is prefix + scope + 0x1f + key, matching the TypeScript store.
func (s *Store) key(scope, k string) string {
	return s.prefix + scope + "\x1f" + k
}

// toString accepts the two shapes a RESP reply element arrives in: a Go string for a bulk string, or a byte
// slice when the client is configured for raw bytes.
func toString(v interface{}) (string, error) {
	switch val := v.(type) {
	case string:
		return val, nil
	case []byte:
		return string(val), nil
	default:
		return "", fmt.Errorf("redis: unexpected reply element type %T", v)
	}
}

// toInt64 accepts a Lua number, which go-redis surfaces as int64, or a Lua string carrying a hash field value
// verbatim.
func toInt64(v interface{}) (int64, error) {
	switch val := v.(type) {
	case int64:
		return val, nil
	case string:
		n, err := strconv.ParseInt(val, 10, 64)
		if err != nil {
			return 0, fmt.Errorf("redis: not an integer: %q", val)
		}
		return n, nil
	default:
		return 0, fmt.Errorf("redis: unexpected reply element type %T", v)
	}
}

// pairsToFields flattens a Lua HGETALL reply (alternating field, value) into a map.
func pairsToFields(pairs []interface{}) (map[string]string, error) {
	fields := make(map[string]string, len(pairs)/2)
	for i := 0; i+1 < len(pairs); i += 2 {
		k, err := toString(pairs[i])
		if err != nil {
			return nil, err
		}
		v, err := toString(pairs[i+1])
		if err != nil {
			return nil, err
		}
		fields[k] = v
	}
	return fields, nil
}

// numField returns the field parsed as an int64, or 0 when absent or not numeric, mirroring the TypeScript
// store's Number(fields.x ?? 0).
func numField(fields map[string]string, name string) int64 {
	v, ok := fields[name]
	if !ok || v == "" {
		return 0
	}
	n, err := strconv.ParseInt(v, 10, 64)
	if err != nil {
		return 0
	}
	return n
}

// fieldsToRow turns a hash's fields into a rowcodec.Row, base64-decoding the body.
func fieldsToRow(scope, key string, fields map[string]string) (rowcodec.Row, error) {
	row := rowcodec.Row{
		Scope:         scope,
		Key:           key,
		Fingerprint:   fields["fingerprint"],
		State:         fields["state"],
		Fence:         numField(fields, "fence"),
		LeaseUntil:    numField(fields, "lease_until"),
		CreatedAt:     numField(fields, "created_at"),
		ExpiresAt:     numField(fields, "expires_at"),
		ResultOmitted: numField(fields, "result_omitted"),
	}
	if row.State == "" {
		row.State = inFlight
	}
	if meta, ok := fields["result_meta"]; ok {
		row.ResultMeta = &meta
	}
	if body, ok := fields["result_body"]; ok {
		decoded, err := base64.StdEncoding.DecodeString(body)
		if err != nil {
			return rowcodec.Row{}, fmt.Errorf("redis: decode result_body: %w", err)
		}
		row.ResultBody = decoded
	}
	return row, nil
}

// Begin runs beginLua: KEYS[1] the hash, ARGV fingerprint, now, leaseMs, ttlMs, graceMs.
func (s *Store) Begin(ctx context.Context, op anyonce.Operation, opts anyonce.BeginOptions) (anyonce.BeginOutcome, error) {
	reply, err := s.beginScript.Run(ctx, s.client,
		[]string{s.key(op.Scope, op.Key)},
		op.Fingerprint,
		strconv.FormatInt(opts.Now.UnixMilli(), 10),
		strconv.FormatInt(opts.Lease.Milliseconds(), 10),
		strconv.FormatInt(opts.TTL.Milliseconds(), 10),
		strconv.FormatInt(s.grace.Milliseconds(), 10),
	).Slice()
	if err != nil {
		return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: %w", err)
	}
	if len(reply) == 0 {
		return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: empty reply")
	}
	tag, err := toString(reply[0])
	if err != nil {
		return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: %w", err)
	}
	switch tag {
	case "acquired":
		fence, err := toInt64(reply[1])
		if err != nil {
			return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: %w", err)
		}
		return anyonce.BeginOutcome{Kind: anyonce.BeginAcquired, Fence: fence}, nil
	case "in_flight":
		leaseUntil, err := toInt64(reply[1])
		if err != nil {
			return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: %w", err)
		}
		return anyonce.BeginOutcome{Kind: anyonce.BeginInFlight, LeaseUntil: time.UnixMilli(leaseUntil).UTC()}, nil
	case "completed", "mismatch":
		fields, err := pairsToFields(reply[1:])
		if err != nil {
			return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: %w", err)
		}
		row, err := fieldsToRow(op.Scope, op.Key, fields)
		if err != nil {
			return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: %w", err)
		}
		record := rowcodec.ToRecord(row)
		kind := anyonce.BeginCompleted
		if tag == "mismatch" {
			kind = anyonce.BeginMismatch
		}
		return anyonce.BeginOutcome{Kind: kind, Record: &record}, nil
	default:
		return anyonce.BeginOutcome{}, fmt.Errorf("redis: begin: unexpected reply tag %q", tag)
	}
}

// Complete runs completeLua: ARGV fence, now, meta, body ('-' for none), omitted.
func (s *Store) Complete(ctx context.Context, op anyonce.Operation, fence int64, result anyonce.StoredResult, now time.Time) (anyonce.CompleteStatus, error) {
	meta, err := rowcodec.EncodeMeta(result)
	if err != nil {
		return "", fmt.Errorf("redis: complete: %w", err)
	}
	body := "-"
	if !result.Omitted && result.Body != nil {
		body = base64.StdEncoding.EncodeToString(result.Body)
	}
	omitted := "0"
	if result.Omitted {
		omitted = "1"
	}
	reply, err := s.completeScript.Run(ctx, s.client,
		[]string{s.key(op.Scope, op.Key)},
		strconv.FormatInt(fence, 10),
		strconv.FormatInt(now.UnixMilli(), 10),
		meta,
		body,
		omitted,
	).Text()
	if err != nil {
		return "", fmt.Errorf("redis: complete: %w", err)
	}
	return anyonce.CompleteStatus(reply), nil
}

// Abandon runs abandonLua: ARGV fence.
func (s *Store) Abandon(ctx context.Context, op anyonce.Operation, fence int64) (anyonce.CompleteStatus, error) {
	reply, err := s.abandonScript.Run(ctx, s.client,
		[]string{s.key(op.Scope, op.Key)},
		strconv.FormatInt(fence, 10),
	).Text()
	if err != nil {
		return "", fmt.Errorf("redis: abandon: %w", err)
	}
	return anyonce.CompleteStatus(reply), nil
}

// Get is an HGETALL plus the logical expiry check; a hash with no fingerprint field counts as absent.
func (s *Store) Get(ctx context.Context, scope, key string, now time.Time) (*anyonce.Record, error) {
	fields, err := s.client.HGetAll(ctx, s.key(scope, key)).Result()
	if err != nil {
		return nil, fmt.Errorf("redis: get: %w", err)
	}
	if _, ok := fields["fingerprint"]; !ok {
		return nil, nil
	}
	row, err := fieldsToRow(scope, key, fields)
	if err != nil {
		return nil, fmt.Errorf("redis: get: %w", err)
	}
	if row.ExpiresAt <= now.UnixMilli() {
		return nil, nil
	}
	record := rowcodec.ToRecord(row)
	return &record, nil
}

// Purge is a no-op that returns 0: PEXPIRE sweeps expired hashes (REQ-ST-REDIS-1, Q21).
func (s *Store) Purge(context.Context, time.Time) (int, error) {
	return 0, nil
}

// PhysicallyRemove deletes the hash unconditionally, which is what a PEXPIRE sweep does. Test-only.
func (s *Store) PhysicallyRemove(ctx context.Context, scope, key string) error {
	if err := s.client.Del(ctx, s.key(scope, key)).Err(); err != nil {
		return fmt.Errorf("redis: physically remove: %w", err)
	}
	return nil
}

File: packages/stores/src/redis.ts
import type {
  BeginOptions,
  BeginOutcome,
  CompleteStatus,
  IdempotencyRecord,
  OmittedResult,
  Operation,
  Store,
  StoredResult,
} from '@anyonce/core';
import { isOmitted } from '@anyonce/core';
import {
  base64ToBytes,
  bytesToBase64,
  encodeResultMeta,
  type RecordRow,
  rowToRecord,
} from './codec';

/** The four calls the store needs; three client shapes are wrapped below. */
export interface RedisAdapter {
  evalsha(sha: string, keys: string[], args: string[]): Promise<unknown>;
  eval(script: string, keys: string[], args: string[]): Promise<unknown>;
  hgetall(key: string): Promise<Record<string, string>>;
  del(key: string): Promise<unknown>;
}

interface IoredisLike {
  evalsha(sha: string, numkeys: number, ...args: string[]): Promise<unknown>;
  eval(script: string, numkeys: number, ...args: string[]): Promise<unknown>;
  hgetall(key: string): Promise<Record<string, string>>;
  del(key: string): Promise<unknown>;
}
interface NodeRedisLike {
  evalSha(sha: string, options: { keys: string[]; arguments: string[] }): Promise<unknown>;
  eval(script: string, options: { keys: string[]; arguments: string[] }): Promise<unknown>;
  hGetAll(key: string): Promise<Record<string, string>>;
  del(key: string): Promise<unknown>;
}
interface UpstashLike {
  evalsha(sha: string, keys: string[], args: string[]): Promise<unknown>;
  eval(script: string, keys: string[], args: string[]): Promise<unknown>;
  hgetall(key: string): Promise<Record<string, unknown> | null>;
  del(key: string): Promise<unknown>;
}

export function fromIoredis(client: IoredisLike): RedisAdapter {
  return {
    evalsha: (sha, keys, args) => client.evalsha(sha, keys.length, ...keys, ...args),
    eval: (script, keys, args) => client.eval(script, keys.length, ...keys, ...args),
    hgetall: (key) => client.hgetall(key),
    del: (key) => client.del(key),
  };
}

export function fromNodeRedis(client: NodeRedisLike): RedisAdapter {
  return {
    evalsha: (sha, keys, args) => client.evalSha(sha, { keys, arguments: args }),
    eval: (script, keys, args) => client.eval(script, { keys, arguments: args }),
    hgetall: (key) => client.hGetAll(key),
    del: (key) => client.del(key),
  };
}

export function fromUpstash(client: UpstashLike): RedisAdapter {
  return {
    evalsha: (sha, keys, args) => client.evalsha(sha, keys, args),
    eval: (script, keys, args) => client.eval(script, keys, args),
    hgetall: async (key) => {
      const out = await client.hgetall(key);
      const flat: Record<string, string> = {};
      for (const [k, v] of Object.entries(out ?? {})) flat[k] = String(v);
      return flat;
    },
    del: (key) => client.del(key),
  };
}

/** KEYS[1] hash; ARGV fingerprint, now, leaseMs, ttlMs, graceMs. Returns {'acquired', fence} | {'in_flight', leaseUntil} | {'completed', ...HGETALL} | {'mismatch', ...HGETALL}. */
export const BEGIN_LUA = `
local k = KEYS[1]
local fp = ARGV[1]
local now = tonumber(ARGV[2])
local lease = tonumber(ARGV[3])
local ttl = tonumber(ARGV[4])
local grace = tonumber(ARGV[5])
local row = redis.call('HMGET', k, 'fingerprint', 'state', 'fence', 'lease_until', 'expires_at')
local fence = 1
if row[1] then
  if tonumber(row[5]) > now then
    if row[1] ~= fp then
      local all = redis.call('HGETALL', k)
      table.insert(all, 1, 'mismatch')
      return all
    end
    if row[2] == 'completed' then
      local all = redis.call('HGETALL', k)
      table.insert(all, 1, 'completed')
      return all
    end
    if tonumber(row[4]) > now then
      return {'in_flight', row[4]}
    end
  end
  fence = tonumber(row[3]) + 1
end
redis.call('HSET', k, 'fingerprint', fp, 'state', 'in_flight', 'fence', fence, 'lease_until', now + lease, 'created_at', now, 'expires_at', now + ttl, 'result_omitted', 0)
redis.call('HDEL', k, 'result_meta', 'result_body')
redis.call('PEXPIRE', k, ttl + grace)
return {'acquired', fence}
`;

/** ARGV fence, now, meta, body ('-' for none), omitted. Returns ok | stale_fence | not_found. */
export const COMPLETE_LUA = `
local k = KEYS[1]
local fence = tonumber(ARGV[1])
local now = tonumber(ARGV[2])
local row = redis.call('HMGET', k, 'fence', 'state', 'expires_at')
if not row[1] or tonumber(row[3]) <= now then return 'not_found' end
if tonumber(row[1]) ~= fence then return 'stale_fence' end
if row[2] == 'completed' then return 'ok' end
redis.call('HSET', k, 'state', 'completed', 'result_meta', ARGV[3], 'result_omitted', ARGV[5])
if ARGV[4] == '-' then redis.call('HDEL', k, 'result_body') else redis.call('HSET', k, 'result_body', ARGV[4]) end
return 'ok'
`;

/** ARGV fence. Returns ok | stale_fence | not_found. */
export const ABANDON_LUA = `
local k = KEYS[1]
local fence = tonumber(ARGV[1])
local row = redis.call('HMGET', k, 'fence', 'state')
if not row[1] or row[2] ~= 'in_flight' then return 'not_found' end
if tonumber(row[1]) ~= fence then return 'stale_fence' end
redis.call('DEL', k)
return 'ok'
`;

export interface RedisStoreOptions {
  adapter: RedisAdapter;
  /** Key prefix. Default anyonce: */
  prefix?: string;
  /** Added to PEXPIRE so a late complete from the previous fence holder still finds its hash. Default 60000. */
  nativeTtlGraceMs?: number;
}

async function sha1Hex(text: string): Promise<string> {
  const digest = new Uint8Array(
    await crypto.subtle.digest('SHA-1', new TextEncoder().encode(text)),
  );
  let out = '';
  for (const b of digest) out += b.toString(16).padStart(2, '0');
  return out;
}

function hashToRow(scope: string, key: string, fields: Record<string, string>): RecordRow {
  return {
    scope,
    key,
    fingerprint: fields.fingerprint ?? '',
    state: (fields.state as RecordRow['state']) ?? 'in_flight',
    fence: Number(fields.fence ?? 0),
    lease_until: Number(fields.lease_until ?? 0),
    created_at: Number(fields.created_at ?? 0),
    expires_at: Number(fields.expires_at ?? 0),
    result_meta: fields.result_meta ?? null,
    result_body: fields.result_body === undefined ? null : base64ToBytes(fields.result_body),
    result_omitted: Number(fields.result_omitted ?? 0),
  };
}

function pairsToFields(pairs: unknown[]): Record<string, string> {
  const out: Record<string, string> = {};
  for (let i = 0; i + 1 < pairs.length; i += 2) out[String(pairs[i])] = String(pairs[i + 1]);
  return out;
}

/** REQ-ST-REDIS-1: every state transition is one Lua script (EVALSHA, EVAL on NOSCRIPT), one hash per record. */
export class RedisStore implements Store {
  private readonly adapter: RedisAdapter;
  private readonly prefix: string;
  private readonly grace: number;
  private readonly shas = new Map<string, string>();

  constructor(options: RedisStoreOptions) {
    this.adapter = options.adapter;
    this.prefix = options.prefix ?? 'anyonce:';
    this.grace = options.nativeTtlGraceMs ?? 60_000;
  }

  keyFor(op: Pick<Operation, 'scope' | 'key'>): string {
    return `${this.prefix}${op.scope}\x1f${op.key}`;
  }

  private async run(script: string, keys: string[], args: string[]): Promise<unknown> {
    let sha = this.shas.get(script);
    if (sha === undefined) {
      sha = await sha1Hex(script);
      this.shas.set(script, sha);
    }
    try {
      return await this.adapter.evalsha(sha, keys, args);
    } catch (error) {
      if (!(error instanceof Error && /NOSCRIPT/i.test(error.message))) throw error;
      return this.adapter.eval(script, keys, args);
    }
  }

  async begin(op: Operation, opts: BeginOptions): Promise<BeginOutcome> {
    const reply = (await this.run(
      BEGIN_LUA,
      [this.keyFor(op)],
      [
        op.fingerprint,
        String(opts.now),
        String(opts.leaseMs),
        String(opts.ttlMs),
        String(this.grace),
      ],
    )) as unknown[];
    const tag = String(reply[0]);
    if (tag === 'acquired') return { outcome: 'acquired', fence: Number(reply[1]) };
    if (tag === 'in_flight') return { outcome: 'in_flight', leaseUntil: Number(reply[1]) };
    const record = rowToRecord(hashToRow(op.scope, op.key, pairsToFields(reply.slice(1))));
    return tag === 'completed' ? { outcome: 'completed', record } : { outcome: 'mismatch', record };
  }

  async complete(
    op: Operation,
    fence: number,
    result: StoredResult | OmittedResult,
    now: number,
  ): Promise<CompleteStatus> {
    const omitted = isOmitted(result);
    const body = omitted || result.body === undefined ? '-' : bytesToBase64(result.body);
    const reply = await this.run(
      COMPLETE_LUA,
      [this.keyFor(op)],
      [String(fence), String(now), encodeResultMeta(result), body, omitted ? '1' : '0'],
    );
    return String(reply) as CompleteStatus;
  }

  async abandon(op: Operation, fence: number): Promise<CompleteStatus> {
    return String(
      await this.run(ABANDON_LUA, [this.keyFor(op)], [String(fence)]),
    ) as CompleteStatus;
  }

  async get(op: Pick<Operation, 'scope' | 'key'>, now: number): Promise<IdempotencyRecord | null> {
    const fields = await this.adapter.hgetall(this.keyFor(op));
    if (fields.fingerprint === undefined) return null;
    const row = hashToRow(op.scope, op.key, fields);
    return row.expires_at > now ? rowToRecord(row) : null;
  }

  /** PEXPIRE sweeps expired hashes; nothing to do here (Q21). */
  async purge(_now: number): Promise<number> {
    return 0;
  }

  /** Test-only: what the native expiry would do. */
  async physicallyRemove(op: Pick<Operation, 'scope' | 'key'>): Promise<void> {
    await this.adapter.del(this.keyFor(op));
  }
}
